syntax = "proto3";

package kmeans;


//  Defining service name along with the function with ReqArg and Response name
//  need multiple services for Master , mapper , reducer

service Master{
  // rpc Master_Mapper (MasterMapperArgs) returns (MasterMapperReply) {}
  // rpc Master_Reducer (MasterReducerArgs) returns (MasterReducerReply) {}
}

service Mapper{
  rpc Mapper (MapperArgs) returns (MapperReply) {}
  rpc PartitionReq (PartitionReqArgs) returns (PartitionReqReply) {}
}

service Reducer{
  rpc Reducer (ReducerArgs) returns (ReducerReply) {}
}


message MapperArgs {
  int32 mapper_id = 1;
  int32 start_index = 2;
  int32 end_index = 3;
  repeated Centroid centroids = 4;
  int32 num_reducers = 5;
}

message Centroid {
  repeated double centroid = 1;
}

message MapperReply {
  string status = 1;
}

message ReducerArgs {
  int32 reducer_id = 1;
  repeated string mapper_addresses = 2;
  int32 num_centroids = 3;

}

message ReducerReply {
  string status = 1;
  repeated computedCentroid computed_centroids = 2;
}
message computedCentroid {
  int32 centroid_key = 1;
  repeated double centroid = 2;
}


message PartitionReqArgs {
  int32 reducer_id = 1;
}

message PartitionReqReply {
  repeated string partition_file_content = 1;
}